aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h

```diff
@@ -0,0 +1,522 @@
+#pragma once
+
+#include <ATen/Tensor.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/Dispatch.h>
+#include <ATen/native/sparse/Macros.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/SparseTensorUtils.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/arange.h>
+#include <ATen/ops/empty.h>
+#include <ATen/ops/_sparse_coo_tensor_with_dims_and_tensors.h>
+#include <ATen/ops/from_blob.h>
+#include <ATen/ops/result_type.h>
+#endif
+
+#ifdef GPUCC
+#define NAME "sparse_binary_op_intersection_cuda"
+#else
+#define NAME "sparse_binary_op_intersection_cpu"
+#endif
+
+#define CALL(...) __VA_ARGS__();
+#define EXPAND(b, n, ...)         \
+  if (b) {                        \
+    using index_t ## n = int32_t; \
+    __VA_ARGS__                   \
+  }                               \
+  else {                          \
+    using index_t ## n = int64_t; \
+    __VA_ARGS__                   \
+  }
+#define BOOL_TO_INDEX_TYPE1(b0, ...) \
+  EXPAND(b0, 0, CALL(__VA_ARGS__))
+#define BOOL_TO_INDEX_TYPE2(b1, b0, ...) \
+  EXPAND(b1, 1, BOOL_TO_INDEX_TYPE1(b0, __VA_ARGS__))
+#define BOOL_TO_INDEX_TYPE3(b2, b1, b0, ...) \
+  EXPAND(b2, 2, BOOL_TO_INDEX_TYPE2(b1, b0, __VA_ARGS__))
+
+namespace at {
+namespace native {
+
+namespace {
+
+using at::sparse::get_sparse_impl;
+
+// ForwardIt: only legacy random access iterator is supported.
+template<class ForwardIt, class T, bool is_lower = true>
+static FUNCAPI INLINE
+ForwardIt find_bound(ForwardIt first, ForwardIt last, const T& value) {
+    ForwardIt RESTRICT it;
+    typename std::iterator_traits<ForwardIt>::difference_type count, step;
+    // NOTE: std::distance(first, last) compiles but produces wrong results on CUDA,
+    // so only legacy random access iterators are safe in this code.
+    count = last - first;
+
+    while (count > 0) {
+      it = first;
+      step = count / 2;
+      // avoiding std::advance(it, step),
+      // although it does work unlike std::distance on CUDA.
+      it += step;
+      if (is_lower ? *it < value : value >= *it) {
+        first = ++it;
+        count -= step + 1;
+      }
+      else {
+        count = step;
+      }
+    }
+    return first;
+}
+
+template <template <typename func_t> class kernel_t>
+struct KernelLauncher {
+  template <typename func_t>
+  static void launch(TensorIteratorBase& iter, const func_t& f) {
+    kernel_t<func_t>::launch(iter, f);
+  }
+};
+
+template <
+  template <typename func_t> class kernel_t,
+  typename binary_op_t,
+  typename index_t = int64_t,
+  typename hash_t = int64_t,
+  typename offset_t = int64_t>
+void _sparse_binary_op_intersection_kernel_impl(
+    Tensor& res,
+    const Tensor& x_,
+    const Tensor& y_,
+    const std::vector<int64_t> broadcasted_shape,
+    const bool is_commutative = true
```

Comment:
Bad name, will rename to `commutes_with_sum`.

Reply:
Agreed, that's a better name